riscv-page-tables/src/page_table.rs

```diff
@@ -1207,6 +1269,137 @@ impl<T: PagingMode> GuestStagePageTable<T> {
+                paddr,
+                page_size,
+                owner,
+                MemType::Ram,
+                prev_page_state,
+                tlb_version,
+            ) {
+                return Err(Error::PageNotPromotable);
+            }
+            prev_page_state = Some(page_state);
+
+            // Move to next page
+            va = Self::check_and_increment_vaddr(va, end, page_size)?;
+        }
+
+        // Promotes the non-leaf PTE and retrieves the now-unused page-table page
+        let (leaf, free_page) = inner.promote_leaf(vaddr.into(), requested_page_size)?;
+
+        Ok((leaf.page_addr(), leaf.level().leaf_page_size(), free_page))
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    /// Verifies the virtual address range is covering only one invalidated page that matches the
+    /// expected page size. Also checks the page state is either "Blocked" and uniquely owned by
+    /// the TVM, or "BlockedShared" and not owned by the TVM. The page must have been invalidated
+    /// with a TLB version that's older than the current one.
+    pub fn demote_page(
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        requested_page_size: PageSize,
+        expected_page_size: PageSize,
+        page_tracker: PageTracker,
+        owner: PageOwnerId,
+        tlb_version: TlbVersion,
+        free_page: Page<InternalClean>,
+    ) -> Result<(SupervisorPageAddr, PageSize)> {
+        vaddr
+            .checked_add_pages_with_size(1, expected_page_size)
+            .ok_or(Error::AddressOverflow)?;
+        let mut inner = self.inner.lock();
+        let pte = inner.get_invalidated_leaf(vaddr)?;
+        let paddr = pte.page_addr();
+        let page_size = pte.level().leaf_page_size();
+        // Check the page matches the expected page size
+        if page_size != expected_page_size {
+            return Err(Error::InvalidPageSize);
+        }
+        // Check the page state
+        if !page_tracker.is_demotable_page(paddr, page_size, owner, MemType::Ram, tlb_version) {
+            return Err(Error::PageNotPromotable);
+        }
+        // Installs a non-leaf PTE pointing the newly-allocated page-table page in place of the huge-page leaf PTE
+        // Safety: `page.addr()` references a page uniquely owned by the root `GuestStagePageTable`.
+        let table_entry = unsafe { pte.map_table(free_page.addr()) };
+        let mut page_table = table_entry.table();
+        // Fills in the free page-table page mapping the sub-pages of the huge page to be split,
+        // marking each PTE as present.
+        for va in vaddr
+            .iter_from_with_size(requested_page_size)
+            .unwrap()
+            .take(512)
+        {
+            let entry = page_table.entry_for_addr_mut(va.into());
+            use TableEntryType::*;
+            let u = match entry {
+                Unused(u) => u,
+                _ => return Err(Error::PteNotUnused),
```

Comment:
should be `unreachable()!` since the page table must've been zeroed out.

Reply:
Agreed.